File: project/jimmer-apt/src/main/java/org/babyfish/jimmer/apt/client/ClientExceptionMetadata.java
package org.babyfish.jimmer.apt.client;

import javax.lang.model.element.TypeElement;
import java.util.List;

public class ClientExceptionMetadata {

    private final TypeElement element;

    private final String family;

    private final String code;

    private final ClientExceptionMetadata superMetadata;

    private List<ClientExceptionMetadata> subMetdatas;

    public ClientExceptionMetadata(TypeElement element, String family, String code, ClientExceptionMetadata superMetadata) {
        this.element = element;
        this.family = family;
        this.code = code;
        this.superMetadata = superMetadata;
    }

    public TypeElement getElement() {
        return element;
    }

    public String getFamily() {
        return family;
    }

    public String getCode() {
        return code;
    }

    public ClientExceptionMetadata getSuperMetadata() {
        return superMetadata;
    }

    public List<ClientExceptionMetadata> getSubMetdatas() {
        return subMetdatas;
    }

    void setSubMetdatas(List<ClientExceptionMetadata> subMetdatas) {
        this.subMetdatas = subMetdatas;
    }
}

File: project/jimmer-ksp/src/main/kotlin/org/babyfish/jimmer/ksp/client/ClientExceptionMetadata.kt
package org.babyfish.jimmer.ksp.client

import com.google.devtools.ksp.symbol.KSClassDeclaration

data class ClientExceptionMetadata(
    val declaration: KSClassDeclaration,
    val family: String,
    val code: String?,
    val superMetadata: ClientExceptionMetadata?
) {
    private lateinit var _subMetadatas: List<ClientExceptionMetadata>

    var subMetadatas: List<ClientExceptionMetadata>
        get() = _subMetadatas
        internal set(value) { _subMetadatas = value }
}